perform exit:
continue even though exit was called...
